piksi_tools/console/settings_view.py

```diff
@@ -601,66 +644,72 @@ def _settings_import_from_file_button_fired(self):
+        parser = configparser.ConfigParser()
+        # the optionxform is needed to handle case sensitive settings
+        parser.optionxform = str
+        try:
+            with open(file.path, 'r') as f:
+                parser.read_file(f)
+        except configparser.ParsingError as e:  # file formatted incorrectly
+            print('Unable to parse ini file due to ParsingError: {}.'.format(e))
+            print('Unable to import settings to device.')
+            return
+        except IOError as e:  # IOError (likely a file permission issue)
+            print('Unable to read ini file due to IOError: {}'.format(e))
+            print('Unable to import settings to device.')
+            return
+
+        # Iterate over each setting
+        for section, settings in parser.items():
+
+            if not settings:
+                # Empty
+                continue
+
+            this_section = self.settings.get(section, None)
+
+            if this_section is None:
+                self._import_failure_section(section)
                 return
 
-            # Iterate over each setting and set in the GUI.
-            # Use the same mechanism as GUI to do settings write to device
-
-            for section, settings in parser.items():
-                this_section = self.settings.get(section, None)
-                for setting, value in settings.items():
-                    if this_section:
-                        this_setting = this_section.get(setting, None)
-                        if this_setting:
-                            this_setting.value = value
-                        else:
-                            print(("Unable to import settings from file. Setting \"{0}\" in section \"{1}\""
-                                   " has not been sent from device.").format(setting, section))
-                            return
-                    else:
-                        print(("Unable to import settings from file."
-                               " Setting section \"{0}\" has not been sent from device.").format(section))
-                        return
-
-            # Double check that no settings had a write failure.  All settings should exist if we get to this point.
-            a = TimedDelayStoppableThread(SETTINGS_REVERT_TIMEOUT + 0.1,
-                                          target=self._wait_for_any_write_failures,
-                                          args=(dict(parser)))
-            a.start()
-        else:
-            pass  # No error message here because user likely pressed cancel when choosing file
+            for setting, value in settings.items():
+                this_setting = this_section.get(setting, None)
+                if this_setting is None:
+                    self._import_failure_not_found(section, setting)
+                    return
+
+                if this_setting.value == value:
+                    continue
```

Comment:
Not writing a value here relies heavily on the console's idea of what a current setting's value is, which has no basis in reality and is basically just a cache.  If we are going to ignore lines based upon the console's cache, we should require a settings.read_all() immediately before importing a file or have some other way to ensure that the cache is updated for the currently connected device before choosing ignore a settings write.  If I ask to write a bunch of settings from a file, I kind of expect them to be written and I don't want the console to outsmart me.

A use case that I have  instructed customers to use to configure devices in bulk is to use the same console session to configure many devices by unplugging and replugging serial into each device one by one.  In that use case, this will be broken because the console will ignore the file based upon its cache of settings values.  Similarly, If a device has restarted, but the console did not receive the STARTUP_MSG triggering a refresh of the cache, the console could ignore a settings write based upon the settings value at the last power-up.

What is the harm in writing a setting even if the value is the same?

Reply:
This also kind of masks the fact that reading settings from a file has taken a pretty big performance hit compared with the old version.  The reason I did not re-use the GUI logic for changing a setting when reading from a file was so you could import an entire file in seconds and verify all the settings in parallel.  I went through and changed the value of every single setting on a Piksi Multi and see that it takes 30 seconds or more in some cases to import that file.  When it takes a long time to import the file, the UI experience isn't very good.  As a user, you are completely unaware that anything is happening until the pop-up appears. 

For a reproduction scenario, over serial at 115200 baud, change all of the settings in the NMEA group to "0", then import a config.ini file with the NMEA settings group set to the default value.  These 8 settings alone take 10 seconds on my mac before the success pop-up.  Imagine a larger file import, the time could get quite long indeed.

If we are going to incur this big performance hit, I almost want the GUI to block until the settings import is complete. Is there some way to make things faster to get closer to the old implementation in performance?

Reply:
Agreed that the GUI thread is not good place to do the writes. Actually I was surprised that this didn't go through the `WorkQueue`.

Reply:
There's `old != new` guard also in the original `_value_changed()` so I don't think the functionality has changed in that regard.

Reply:
I'll remove all the `old` vs `new` guard.